### PR TITLE
Clear cookies with optional url

### PR DESCRIPTION
**Pull request by graefenhain**

### CODE REVIEW

File: ios/Plugin/Plugin.swift [hunk 1]
```diff
@@ -130,14 +130,29 @@ import Foundation
...
+            }
+            jar.cookies(for: url)?.forEach { jar.deleteCookie($0) }
+        } else {
+            self.deleteAllCookies()
+        }
         
         call.resolve()
     }
+    
+    private func deleteAllCookies() {
```
Comment (shanselm-ergon) [inline ios/Plugin/Plugin.swift:147]: won't compile, `call` and `jar` are not defined.

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 2]
```diff
@@ -259,7 +260,17 @@ public void deleteCookie(PluginCall call) {
 
     @PluginMethod
     public void clearCookies(PluginCall call) {
-        cookieManager.removeAllCookies();
+        String urlString = call.getString("url");
+        if (urlString == null) {
+            cookieManager.removeAllCookies();
+        } else {
+            try {
+                URL url = new URL(urlString);
```
Comment (7-plus-t) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:268]: Why transform it to an URL which is then not used?

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 3]
```diff
@@ -259,7 +260,17 @@ public void deleteCookie(PluginCall call) {
...
+                cookieManager.removeAllCookiesFor(urlString);
+            } catch (MalformedURLException e) {
+                call.reject("Could not parse URL. Check that url has valid format");
```
Comment (7-plus-t) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:271]: Typo: `url` -> `URL`

File: android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java [hunk 4]
```diff
@@ -139,6 +140,23 @@ public void removeAllCookies() {
...
+    /**
+     * Removes all cookies for the given url. This method is asynchronous.
+     *
+     * @param urlString
+     */
+    public void removeAllCookiesFor(String urlString) {
+        String cookieString = webkitCookieManager.getCookie(urlString);
+        /**
+         * TODO:
+         * The Webkit CookieManager API has not method to delete a single cookie - looks as if
```
Comment (7-plus-t) [inline android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java:152]: Typo: `has not` -> `has no`

File: android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java [hunk 5]
```diff
@@ -139,6 +140,23 @@ public void removeAllCookies() {
         flush();
     }
 
+    /**
+     * Removes all cookies for the given url. This method is asynchronous.
+     *
+     * @param urlString
+     */
+    public void removeAllCookiesFor(String urlString) {
+        String cookieString = webkitCookieManager.getCookie(urlString);
```
Comment (7-plus-t) [inline android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java:149]: unused

File: android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java [hunk 6]
```diff
@@ -139,6 +140,23 @@ public void removeAllCookies() {
         flush();
     }
 
+    /**
+     * Removes all cookies for the given url. This method is asynchronous.
+     *
+     * @param urlString
+     */
+    public void removeAllCookiesFor(String urlString) {
```
Comment (7-plus-t) [inline android/src/main/java/com/getcapacitor/plugin/http/CapacitorCookieManager.java:148]: Method named with Swift naming convention (`For` only makes sense having parameter labels).

File: ios/Plugin/Plugin.swift [hunk 7]
```diff
@@ -130,14 +130,29 @@ import Foundation
...
+        guard let cookies = jar.cookies else {
+            call.resolve()
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift]: Does not compile. `call` is not available here.

File: ios/Plugin/CapacitorCookieManager.swift [hunk 8]
```diff
@@ -37,7 +37,12 @@ public class CapacitorCookieManager {
         if cookie != nil { jar.deleteCookie(cookie!) }
     }
     
-    public func clearCookies(_ url: URL) {
-        jar.cookies(for: url)?.forEach({ (cookie) in jar.deleteCookie(cookie) })
+    public func clearCookies(_ url: URL?) {
+        if let url = url {
+            jar.cookies(for: url)?.forEach({ (cookie) in jar.deleteCookie(cookie) })
```
Comment (7-plus-t) [inline ios/Plugin/CapacitorCookieManager.swift:42]: Brackets not needed, could write:
`jar.cookies(for: url)?.forEach { jar.deleteCookie($0) }`

File: ios/Plugin/CapacitorCookieManager.swift [hunk 9]
```diff
@@ -37,7 +37,12 @@ public class CapacitorCookieManager {
...
+        } else {
+            let cookies = HTTPCookieStorage.shared.cookies
```
Comment (7-plus-t) [inline ios/Plugin/CapacitorCookieManager.swift:44]: Why not use `jar.cookies` and inline as you did above?

File: ios/Plugin/CapacitorCookieManager.swift [hunk 10]
```diff
@@ -37,7 +37,12 @@ public class CapacitorCookieManager {
...
+            cookies?.forEach { HTTPCookieStorage.shared.deleteCookie($0) }
```
Comment (7-plus-t) [inline ios/Plugin/CapacitorCookieManager.swift:45]: use `jar` instead of `HTTPCookieStorage.shared`

File: ios/Plugin/Plugin.swift [hunk 11]
```diff
@@ -130,14 +130,29 @@ import Foundation
...
 
     @objc func clearCookies(_ call: CAPPluginCall) {
-        guard let url = call.getServerURLOrReject() else { return }
-        
         let jar = HTTPCookieStorage.shared
-        jar.cookies(for: url)?
-            .forEach { jar.deleteCookie($0) }
+        if let urlString = (call.getString("url")) {
+            guard let url = URL(string: urlString) else {
+                call.reject("Could not parse URL. Check that url has valid format")
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:136]: Typo: `url` -> `URL`

File: ios/Plugin/Plugin.swift [hunk 12]
```diff
@@ -130,14 +130,29 @@ import Foundation
     }
 
     @objc func clearCookies(_ call: CAPPluginCall) {
-        guard let url = call.getServerURLOrReject() else { return }
-        
         let jar = HTTPCookieStorage.shared
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:133]: Shouldn't this use cookieManager? Otherwise the clear method is never used.

File: ios/Plugin/Plugin.swift [hunk 13]
```diff
@@ -130,14 +130,29 @@ import Foundation
     }
 
     @objc func clearCookies(_ call: CAPPluginCall) {
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:132]: The API would be cleaner if this method would be split into two methods:
`clearCookies(...)` and `clearAllCookies()`

File: ios/Plugin/Plugin.swift [hunk 14]
```diff
@@ -130,14 +130,29 @@ import Foundation
...
-        
         let jar = HTTPCookieStorage.shared
-        jar.cookies(for: url)?
-            .forEach { jar.deleteCookie($0) }
+        if let urlString = (call.getString("url")) {
+            guard let url = URL(string: urlString) else {
+                call.reject("Could not parse URL. Check that url has valid format")
+                return
+            }
+            jar.cookies(for: url)?.forEach { jar.deleteCookie($0) }
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:139]: Use cookieManager instead

File: ios/Plugin/Plugin.swift (hunk 1 above)
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:147]: Method not needed, as it is required in cookieManager

File: src/web.ts [hunk 15]
```diff
@@ -124,7 +124,7 @@ export class HttpWeb extends WebPlugin implements HttpPlugin {
    */
   public clearCookies = async (
     // @ts-ignore
-    options: HttpMultiCookiesOptions,
+    options?: HttpMultiCookiesOptions,
```
Comment (7-plus-t) [inline src/web.ts:127]: Seems weird, that both, `options` and `url` is optional. Would be fixed when splitting the method.

File: ios/Plugin/CapacitorCookieManager.swift (hunk 10 above)
Comment (shanselm-ergon) [inline ios/Plugin/CapacitorCookieManager.swift:45]: Here you could also use `jar.cookies?`, or `let cookies = jar.cookies`

**Review by 7-plus-t — CHANGES_REQUESTED** [review]

File: src/web.ts (hunk 15 above)
Comment (shanselm-ergon) [reply src/web.ts:127]: No, the options shouldn't be optional anymore, only the URL in the options. But yes, with a split this problem would be obsolet. 
In the definitions.ts it is not obsolet anymore, so this is only a failure in the web implementation. Did this not crash while building?

File: ios/Plugin/Plugin.swift (hunk 7 above)
Comment (graefenhain) [reply ios/Plugin/Plugin.swift]: already fixed

File: ios/Plugin/Plugin.swift (hunk 11 above)
Comment (graefenhain) [reply ios/Plugin/Plugin.swift:136]: you mean `URL` -> `url`, right? Typo in comment

File: ios/Plugin/CapacitorCookieManager.swift (hunk 8 above)
Comment (graefenhain) [reply ios/Plugin/CapacitorCookieManager.swift:42]: already fixed, realised myself

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java (hunk 2 above)
Comment (graefenhain) [reply android/src/main/java/com/getcapacitor/plugin/http/Http.java:268]: To catch the MalformedURLException and then reject the call. But probably there's a better way to check if the URL is valid.

Comment (graefenhain) [reply android/src/main/java/com/getcapacitor/plugin/http/Http.java:268]: extracted to a separate method

File: src/web.ts (hunk 15 above)
Comment (graefenhain) [reply src/web.ts:127]: So splitting the method AND leaving url optional? This is weird as well... Or:
`clearCookies(options: HttpMultiCookiesOptions): Promise<void>;` and
`clearAllCookies(): Promise<void>;` without any options? Is this allowed?

Comment (shanselm-ergon) [reply src/web.ts:127]: yes, the suggestion looks good.

Comment (shanselm-ergon) [inline src/web.ts:127]: this should not be optional (only in the options object we have the url optional)

**Pull request merged**